### PR TITLE
Support relabeling when some event happens

### PR DESCRIPTION
**Pull request by kennytm**
Let Homu automatically relabel the issue when some predefined events happened. The labels can be configured through `cfg.toml` via the `[repo.NAME.labels.EVENT]` key, e.g.

```toml
# when a merge conflict is detected...
[repo.rust.labels.conflict]
# remove the `S-waiting-on-bors` and `S-waiting-on-review` labels...
remove = ['S-waiting-on-bors', 'S-waiting-on-review']
# add the `S-waiting-on-author` label...
add = ['S-waiting-on-author']
# but don't do anything if the PR is already labeled with `S-blocked`, 
# `S-waiting-on-crater` or `S-waiting-on-team`.
except = ['S-blocked', 'S-waiting-on-crater', 'S-waiting-on-team']
```

See `cfg.sample.toml` for a list of supported events.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/homu/141)

### CODE REVIEW

**Comment by kennytm** [comment]
(Potential usage on rust-lang/rust:

```toml
[repo.rust.labels.approved]    # after homu received `r+`
remove = ['S-blocked', 'S-waiting-on-author', 'S-waiting-on-bors', 'S-waiting-on-crater', 'S-waiting-on-review', 'S-waiting-on-team']
add = ['S-waiting-on-bors']

[repo.rust.labels.rejected]    # after homu received `r-`
remove = ['S-blocked', 'S-waiting-on-author', 'S-waiting-on-bors', 'S-waiting-on-crater', 'S-waiting-on-review', 'S-waiting-on-team']
add = ['S-waiting-on-author']

[repo.rust.labels.failed]      # test failed
remove = ['S-blocked', 'S-waiting-on-author', 'S-waiting-on-bors', 'S-waiting-on-crater', 'S-waiting-on-review', 'S-waiting-on-team']
add = ['S-waiting-on-review']

[repo.rust.labels.timed_out]   # test timed out (after 10 hours)
remove = ['S-blocked', 'S-waiting-on-author', 'S-waiting-on-bors', 'S-waiting-on-crater', 'S-waiting-on-review', 'S-waiting-on-team']
add = ['S-waiting-on-review']

[repo.rust.labels.try]         # after homu received `try`
add = ['S-waiting-on-bors']

[repo.rust.labels.try_succeed] # try-build successful
remove = ['S-waiting-on-bors']

[repo.rust.labels.try_failed]  # try-build failed
remove = ['S-waiting-on-bors', 'S-waiting-on-review']
add = ['S-waiting-on-author']

[repo.rust.labels.pushed]      # user pushed a commit after `r+`/`try`
remove = ['S-waiting-on-bors', 'S-waiting-on-author']
add = ['S-waiting-on-review']
except = ['S-blocked', 'S-waiting-on-crater', 'S-waiting-on-team']

[repo.rust.labels.conflict]    # a merge conflict is detected
remove = ['S-waiting-on-bors', 'S-waiting-on-review']
add = ['S-waiting-on-author']
except = ['S-blocked', 'S-waiting-on-crater', 'S-waiting-on-team']
```

)

**Comment by cgwalters** [comment]
I love the idea; we're currently doing this manually; see e.g. the `homu/approved` label on <https://github.com/ostreedev/ostree/pull/1409>.

Your proposed config though is...verbose.  I understand we want it to be configurable but...eek.

Here's one idea I had; labels are good, but what about doing in-place editing of the approval comment?  That feels appropriate for more "detailed" status.  The main thing I personally want *also* as a label is simply whether or not the PR is in the queue.

**Comment by cgwalters** [comment]
To clarify my proposal is that homu have one default label - whether a PR is in queue or not.

Then, breaking down your scheme...first I think we should consider explicitly erroring in the "push after r+" scenario, i.e. the PR gets removed from the queue.  That means the approval flag is dropped and conceptually it goes back to "needs review".

Conflicts...yeah, could use a new label.  But we always remove the in-queue label in this case.

Do we need labels at all for "try"?

For "rejected", that's just removing the approved label.

For test failed/timed out...same deal, just remove approved label by default, but allow users to opt-in to adding labels?

**Comment by cgwalters** [comment]
(Of course I know your PR is from the rust perspective, I have only a single patch in rust.  We have a separate instance of homu for https://github.com/projectatomic/ and so my perspective is mostly colored by that)

**Comment by jdm** [comment]
For Servo we built a system that looks for homu messages and applies or removes labels based on that, so the complexity of Rust's setup is probably similarly useful for us.

**Comment by kennytm** [comment]
We don't actually need the tags for try builds on Rust, but since I don't know how others will use this I shoot for the most general cases here 😄.

File: homu/server.py [hunk 1]
```diff
@@ -328,7 +329,7 @@ def github():
             state = PullReqState(pull_num, head_sha, '', g.db, repo_label,
                                  g.mergeable_que, g.gh,
                                  info['repository']['owner']['login'],
-                                 info['repository']['name'], g.repos)
+                                 info['repository']['name'], {}, g.repos)
```
Comment (jdm) [inline homu/server.py]: Why are we passing an empty dictionary here?

File: cfg.sample.toml [hunk 2]
```diff
@@ -96,6 +96,28 @@ try_users = []
...
+#[repo.NAME.labels.exempted]    # test exempted
+#[repo.NAME.labels.timed_out]   # test timed out (after 10 hours)
+#[repo.NAME.labels.interrupted] # test interrupted (buildbot only)
+#[repo.NAME.labels.try]         # after homu received `try`
+#[repo.NAME.labels.try_succeed] # try-build successful
+#[repo.NAME.labels.try_failed]  # try-build failed
+#[repo.NAME.labels.pushed]      # user pushed a commit after `r+`/`try`
+#remove = ['list', 'of', 'labels', 'to', 'remove']
+#add = ['list', 'of', 'labels', 'to', 'add']
+#except = [
```
Comment (jdm) [inline cfg.sample.toml]: What if we call this `unless` instead?

**Comment by kennytm** [comment]
@jdm Thanks! All fixed.

**Comment by jdm** [comment]
Go ahead and squash the commits.
@bors-servo delegate+

**Pull request merged**